src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+                        "--json=short",
+                        NULL, /* maybe --component=, --root=, or --image= */
+                        NULL, /* maybe --offline */
+                        NULL, /* list, check-new, update, vacuum */
+                        NULL, /* maybe version (for list, update) */
+                        NULL
+                };
+                unsigned k = 1;
+
+                if (setenv("NOTIFY_SOCKET", "/run/systemd/sysupdate/notify", 1) < 0) {
+                        log_error_errno(errno, "setenv() failed: %m");
+                        _exit(EXIT_FAILURE);
+                }
+
+                r = setenv_systemd_exec_pid(true);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to update $SYSTEMD_EXEC_PID, ignoring: %m");
+
+                if (j->target->class == TARGET_HOST)
+                        r = 0; /* Host doesn't need any argument */
+                else if (j->target->class == TARGET_COMPONENT)
+                        r = asprintf(&target_arg, "--component=%s", j->target->name);
+                else if (IN_SET(j->target->image_type, IMAGE_DIRECTORY, IMAGE_SUBVOLUME))
+                        r = asprintf(&target_arg, "--root=%s", j->target->path);
+                else
+                        r = asprintf(&target_arg, "--image=%s", j->target->path);
+                if (r < 0) {
+                        log_oom();
+                        _exit(EXIT_FAILURE);
+                }
+                if (target_arg)
+                        cmd[k++] = target_arg;
+
+                if (j->offline)
+                        cmd[k++] = "--offline";
+
+                switch (j->type) {
+                case JOB_LIST:
+                        cmd[k++] = "list";
+                        break;
+
+                case JOB_DESCRIBE:
+                        cmd[k++] = "list";
+                        assert(!isempty(j->version));
+                        cmd[k++] = j->version;
+                        break;
+
+                case JOB_CHECK_NEW:
+                        cmd[k++] = "check-new";
+                        break;
+
+                case JOB_UPDATE:
+                        cmd[k++] = "update";
+                        cmd[k++] = empty_to_null(j->version);
+                        break;
+
+                case JOB_VACUUM:
+                        cmd[k++] = "vacuum";
+                        break;
+
+                default:
+                        assert_not_reached();
+                }
+
+                execv(sysupdate_binary_path(), (char * const *) cmd);
+                log_error_errno(errno, "Failed to execute systemd-sysupdate: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        r = sd_event_add_child(j->manager->event, &j->pid_event_source, j->pid,
+                               WEXITED, job_on_pid, j);
+        if (r < 0)
+                return r;
+
+        j->stdout_fd = TAKE_FD(stdout_fd);
+        j->pid = TAKE_PID(pid);
```

Comment:
as mentioned elsewhere, please use sd_event_source_set_child_process_own(), and don't store the pid separately